File: planctl/session_markers.py
"""Per-session marker files — the cross-language guard contract.

One JSON file per Claude session at
``~/.local/state/planctl/sessions/<session_id>.json``, schema_version 1:

    {schema_version, session_id, kind: "work"|"close",
     task_id (work) | epic_id (close), created_at}

The TS hook dispatchers (PreToolUse commit deny, SubagentStop worker guard,
Stop checklist guard) read these files to decide whether a main-context agent
is mid-task. Field names and ``kind`` values are a cross-language contract:
deviating from this schema breaks the dispatchers silently.

**Fail open.** The session id comes from ``CLAUDE_CODE_SESSION_ID`` read here
directly — absent env var makes every helper a silent no-op (manual and test
invocations must not error or write). All filesystem errors are swallowed:
marker IO never fails the verb that calls it. A write on a verb's error path
would create a guard lockout for a task that was never claimed, so callers
invoke these strictly on their success path.
"""

from __future__ import annotations

import json
import os
import time
from pathlib import Path

SCHEMA_VERSION = 1

#: Markers older than this are unlinked when a reader touches them.
_STALE_AFTER_SECONDS = 7 * 24 * 60 * 60


def _session_id() -> str | None:
    """Resolve the session id from the env, fail-open.

    Distinct from ``invocation._build``'s fail-CLOSED resolution: an absent
    env var here returns ``None`` (caller no-ops) rather than raising.
    """
    return os.environ.get("CLAUDE_CODE_SESSION_ID") or None


def _sessions_dir() -> Path:
    """The session-marker directory (``~/.local/state/planctl/sessions``).

    Resolved at call time so a monkeypatched ``HOME`` (tests) is honored.
    Mirrors ``run_epic_create``'s ``~/.local/state/planctl/`` precedent.
    """
    return Path("~/.local/state/planctl/sessions").expanduser()


def _marker_path(session_id: str) -> Path:
    return _sessions_dir() / f"{session_id}.json"


def _write_marker(kind: str, id_field: str, target_id: str) -> None:
    """Write the marker for the current session. Silent no-op when no session
    id; all filesystem errors swallowed."""
    session_id = _session_id()
    if session_id is None:
        return
    record = {
        "schema_version": SCHEMA_VERSION,
        "session_id": session_id,
        "kind": kind,
        id_field: target_id,
        "created_at": _now_iso(),
    }
    try:
        path = _marker_path(session_id)
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(json.dumps(record), encoding="utf-8")
    except OSError:
        # Marker IO must never fail the verb.
        return


def _clear_if_matches(id_field: str, target_id: str) -> None:
    """Unlink the current session's marker only when its ``id_field`` matches
    ``target_id``. A mismatched marker is left intact; all errors swallowed."""
    session_id = _session_id()
    if session_id is None:
        return
    try:
        path = _marker_path(session_id)
        raw = path.read_text(encoding="utf-8")
        record = json.loads(raw)
    except (OSError, ValueError):
        return
    if not isinstance(record, dict):
        return
    if record.get(id_field) != target_id:
        return
    try:
        path.unlink()
    except OSError:
        return


def _now_iso() -> str:
    """Local ``now_iso`` so importing this module never pulls in the store."""
    from planctl.store import now_iso

    return now_iso()


def write_work_marker(task_id: str) -> None:
    """Mark this session as actively working ``task_id`` (``kind="work"``).

    Called on the success path of ``claim`` and ``worker resume``.
    """
    _write_marker("work", "task_id", task_id)


def write_close_marker(epic_id: str) -> None:
    """Mark this session as closing ``epic_id`` (``kind="close"``).

    Called on the success path of ``close-preflight``.
    """
    _write_marker("close", "epic_id", epic_id)


def clear_work_marker(task_id: str) -> None:
    """Clear the work marker, but only if it names ``task_id``.

    Called by ``done`` and ``block``: a session whose marker names a different
    task must not have it cleared by a sibling verb.
    """
    _clear_if_matches("task_id", task_id)


def clear_close_marker(epic_id: str) -> None:
    """Clear the close marker, but only if it names ``epic_id``.

    Called by ``close-finalize`` on every terminal outcome.
    """
    _clear_if_matches("epic_id", epic_id)


def read_marker(session_id: str) -> dict | None:
    """Return the parsed marker for ``session_id``, or ``None``.

    Unlinks-and-returns-``None`` for a marker older than 7 days (stale-on-read)
    and for an unparseable / non-dict file. All filesystem errors are swallowed.
    Provided for the guard dispatchers' Python-side tests and any in-process
    reader; the TS dispatchers read the same files independently.
    """
    try:
        path = _marker_path(session_id)
        if not path.exists():
            return None
        if (time.time() - path.stat().st_mtime) > _STALE_AFTER_SECONDS:
            try:
                path.unlink()
            except OSError:
                pass
            return None
        record = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, ValueError):
        return None
    if not isinstance(record, dict):
        try:
            _marker_path(session_id).unlink()
        except OSError:
            pass
        return None
    return record

File: src/session_markers.ts
// Per-session marker files — the writer side of the cross-language guard
// contract (planctl/session_markers.py). claim writes a work marker on its
// success path; done/block clear it (only when it names their task).
//
// One JSON file per session at `~/.local/state/planctl/sessions/<sid>.json`,
// schema_version 1: {schema_version, session_id, kind, task_id|epic_id,
// created_at}. The TS hook dispatchers (plugin/hooks/lib.ts) read these files;
// the field names + `kind` values are the contract.
//
// Fail OPEN: the session id comes from CLAUDE_CODE_SESSION_ID — absent makes
// every helper a silent no-op. All filesystem errors are swallowed: marker IO
// never fails the verb. Callers invoke these strictly on the success path (a
// marker for an unclaimed task would lock out commits).

import {
  existsSync,
  mkdirSync,
  readFileSync,
  unlinkSync,
  writeFileSync,
} from "node:fs";
import { homedir } from "node:os";
import { dirname, join } from "node:path";

import { nowIso } from "./store.ts";

const SCHEMA_VERSION = 1;

/** Resolve the session id from the env, fail-open (empty/absent → null). */
function sessionId(): string | null {
  return process.env.CLAUDE_CODE_SESSION_ID || null;
}

/** `~/.local/state/planctl/sessions` honoring a mutated $HOME (tests). Mirrors
 * lib.ts sessionsDir + the Python expanduser semantics. */
function sessionsDir(): string {
  const home = process.env.HOME || homedir();
  return join(home, ".local", "state", "planctl", "sessions");
}

function markerPath(sid: string): string {
  return join(sessionsDir(), `${sid}.json`);
}

/** Write the marker for the current session. Silent no-op when no session id;
 * all filesystem errors swallowed. */
function writeMarker(kind: string, idField: string, targetId: string): void {
  const sid = sessionId();
  if (sid === null) {
    return;
  }
  const record: Record<string, unknown> = {
    schema_version: SCHEMA_VERSION,
    session_id: sid,
    kind,
    [idField]: targetId,
    created_at: nowIso(),
  };
  try {
    const path = markerPath(sid);
    mkdirSync(dirname(path), { recursive: true });
    writeFileSync(path, JSON.stringify(record));
  } catch {
    // Marker IO must never fail the verb.
  }
}

/** Unlink the current session's marker only when its `idField` matches
 * `targetId`. A mismatched marker is left intact; all errors swallowed. */
function clearIfMatches(idField: string, targetId: string): void {
  const sid = sessionId();
  if (sid === null) {
    return;
  }
  const path = markerPath(sid);
  let record: unknown;
  try {
    if (!existsSync(path)) {
      return;
    }
    record = JSON.parse(readFileSync(path, "utf-8"));
  } catch {
    return;
  }
  if (record === null || typeof record !== "object") {
    return;
  }
  if ((record as Record<string, unknown>)[idField] !== targetId) {
    return;
  }
  try {
    unlinkSync(path);
  } catch {
    // best-effort
  }
}

/** Mark this session as actively working `taskId` (kind="work"). Called on the
 * success path of claim. Mirrors write_work_marker. */
export function writeWorkMarker(taskId: string): void {
  writeMarker("work", "task_id", taskId);
}

/** Clear the work marker, but only if it names `taskId`. Called by done/block.
 * Mirrors clear_work_marker. */
export function clearWorkMarker(taskId: string): void {
  clearIfMatches("task_id", taskId);
}
